crates/fluvio-spu/src/replication/leader/kv.rs

```diff
@@ -0,0 +1,367 @@
+                self.last_seen_offset = record.offset;
+                Poll::Ready(Some(Ok(record.record.into_value().into_vec())))
+            }
+            None => {
+                self.inner_iter_fut = InnerFutureState::None;
+                cx.waker().wake_by_ref();
+                Poll::Pending
+            }
+        }
+    }
+}
+
+enum InnerFutureState {
+    None,
+    Resolved(Vec<RecordItem>),
+    Pending(Pin<Box<dyn Future<Output = Result<Vec<RecordItem>>> + Send + 'static>>),
+}
+
+#[cfg(test)]
+mod tests {
+    use std::{env::temp_dir, path::Path};
+
+    use fluvio_controlplane::replica::Replica;
+    use fluvio_kv_storage::KVStorage;
+    use fluvio_protocol::record::ReplicaKey;
+    use fluvio_storage::{config::ReplicaConfig, FileReplica};
+    use fluvio_types::PartitionId;
+    use flv_util::fixture::ensure_clean_dir;
+    use futures_util::StreamExt;
+
+    use crate::{
+        config::ReplicationConfig, control_plane::StatusMessageSink,
+        storage::SharableReplicaStorage,
+    };
+
+    use super::*;
+
+    #[fluvio_future::test]
+    async fn test_replica_kv() {
+        //given
+        let leader = create_replica("test_replica_kv").await;
+        let notifier = FollowerNotifier::shared();
+        let mut kv = LeaderKVStorage::<String, String, _>::new(LeaderReplicaLog::new(
+            leader.clone(),
+            notifier.clone(),
+        ));
+
+        //when
+        kv.put("key1", "value1").await.unwrap();
+        kv.put("key2", "value2").await.unwrap();
+        kv.put("key3", "value3").await.unwrap();
+
+        kv.delete(&"key2".to_string()).await.unwrap();
+
+        drop(kv);
+
+        let mut kv = LeaderKVStorage::<String, String, _>::new(LeaderReplicaLog::new(
+            leader.clone(),
+            notifier,
+        ));
+        kv.sync_from_log().await.expect("synced");
+
+        //then
+
+        assert_eq!(
+            kv.get(&"key1".to_string()).await.expect("read key1"),
+            Some("value1".to_string())
+        );
+
+        assert_eq!(kv.get(&"key2".to_string()).await.expect("read key1"), None);
+
+        assert_eq!(
+            kv.get(&"key3".to_string()).await.expect("read key1"),
+            Some("value3".to_string())
+        );
+
+        leader.remove().await.expect("removed");
+    }
+
+    #[fluvio_future::test]
+    async fn test_log() {
+        //given
+        let leader = create_replica("test_log").await;
+        let notifier = FollowerNotifier::shared();
+
+        //when
+        let mut log = LeaderReplicaLog::new(leader.clone(), notifier);
+        log.append_batch(vec![b"1".to_vec()]).await.unwrap();
+        log.append_batch(vec![b"2".to_vec(), b"3".to_vec()])
+            .await
+            .unwrap();
+        log.append_batch(vec![b"4".to_vec()]).await.unwrap();
+        let output: Vec<Result<Vec<u8>>> = log.read_from_end().await.expect("read").collect().await;
+
+        //then
+        leader.remove().await.expect("removed");
```

Comment:
dont' understand this here

Reply:
To remove files from disk after test is done